### PR TITLE
osutil: make EnsureDirState remove managed files on failure

### PR DESCRIPTION
**Pull request by zyga**
This patch changes the aforementioned to switch to erase mode when an
error during file-writing stage. This affects how the function is
supposed to be used. Instead of managing all the files in a given
directory or governed by a broad namespace (e.g. "*.snap") the function
is supposed to be used to manage files scoped to a specific snap (e.g.
"docker.*.snap").

This way the erase mode actually helps to attain sanity as to what
happens on errors and errors have isolated impact (other apps are snaps
are not harmed by an error in unrelated snaps)

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: osutil/sync_dir.go [hunk 1]
```diff
@@ -51,10 +52,16 @@ var errSameState = fmt.Errorf("file state has not changed")
...
 // Sub-directories in the name are not allowed.
 //
-// The function stops at the first encountered error but reports all of the
-// changes performed so far. Information about the performed changes is
-// returned to the caller for any extra processing that might be required (e.g.
-// to run some helper program).
+// If writing any of the files fails, EnsureDirState switches to erase mode
+// where *all* of the files managed by the glob pattern are removed (including
+// those that may have been already written). The return value is an empty list
+// of changed files, the real list of removed files and the first error.
```
Comment (niemeyer) [inline osutil/sync_dir.go]: `..., the list of removed files including those that were previously written and then removed, and the first error.`

This behavior also needs to be properly used. Right now profiles are not being unloaded if there are errors.

Comment (zyga) [inline osutil/sync_dir.go]: Yes, that's true. This is the first step towards getting there.

**Comment by stevenwilkin** [comment]
:+1:

File: osutil/sync_dir.go [hunk 2]
```diff
@@ -74,13 +82,21 @@ func EnsureDirState(dir, glob string, content map[string]*FileState) (changed, r
...
+			// writing loop stops. The subsequent erase loop will remove all
+			// the managed content.
+			firstErr = err
+			content = nil
+			changed = nil
+			break
 		}
 		changed = append(changed, baseName)
 	}
 	matches, err := filepath.Glob(filepath.Join(dir, glob))
```
Comment (mvo5) [inline osutil/sync_dir.go]: Maybe a small comment here that the "delete" phase of the code starts here?

File: osutil/sync_dir.go (hunk 1 above)
Comment (zyga) [inline osutil/sync_dir.go]: This is now properly used elsewhere (udev/aaparmor backends)

**Pull request merged**